### DISCUSSION TITLE
Should one use json or jsonl? 100x memory usage when loading json

### DISCUSSION THREAD

**Discussion by RuABraun** [lhotse-speech/lhotse#1220] [Q&A] [answered]
Anyways training with webdatasets so thought it'd be best to use `json` (feature of being ab;e to partially load a `jsonl` not needed) but when I try and load a 5GB `json.gz` cutset it takes 100x the memory. Now wondering whether I should use jsonl instead? I didn't see a suggestion in the docs.

Any tips regarding how to load and write jsons quickly would be appreciated.

**Comment by pzelasko** [comment, marked as answer]
I recommend JSONL, every other format is problematic in some way. We may remove JSON support at some point or make it actually use JSONL.

**Reply by RuABraun** [reply]
Ok thank you.

I do think json is nice just to be able to inspect datasets quickly

**Reply by pzelasko** [reply]
JSONL is the same thing, just one record per line so you don’t have to load it all into memory.

**Reply by RuABraun** [reply]
Are you using some special tool to view a jsonl because I otherwise don't understand how you could say that jsonl is as readable as json? Json is
```
{
  uid: ..
  text: ..
  speaker: ..
}
```
whereas jsonl is 
```
{ uid: ...,  text: .., speaker: ..}
```
If there were always 3 entries this would be fine, but there are more like 30 which makes it impossible to quickly visually scan for certain keys. Maybe your eyesight is just much better than mine

**Reply by pzelasko** [reply]
If you need pretty formatting you can use `jq`, it will even add colors.

**Reply by RuABraun** [reply]
Is there a reasonable way to speed up writing a jsonl ? It's taking several hours for me for a large cutset

**Reply by pzelasko** [reply]
Install orjson if you haven’t already; you can write to stdout („-„) and pipe to gzip so the compression is in another process; otherwise split the manifest into chunks and process in parallel

**Reply by RuABraun** [reply]
It seems to me the main bottleneck is creating a dict for each line. If I use orjson to write it's muuuch faster, but then one can't read them because lhotse expects the "type" field. :(

**Reply by pzelasko** [reply]
If you use `cut.to_dict()` and then write that with orjson it should be ok (it will add `type`).

https://github.com/lhotse-speech/lhotse/blob/d474be6b1d469863064142fc99e0861d164d8311/lhotse/cut/base.py#L203-L205

**Reply by pzelasko** [reply]
(see also `lhotse.serialization.save_to_jsonl`/`CutSet.to_file(".../abc.jsonl[.gz]")`)

**Reply by RuABraun** [reply]
Thanks, that does help, timings for a smaller cutset:
```
to_file() : 80s
to_dict() -> orjson : 50s
orjson: 20s
```
But one can see from the above results that `to_dict()` is an expensive step.

To add some extra info the reason I'm investigating this is because for some datasets I'm using `to_file()` takes many hours.

**Reply by pzelasko** [reply]
Yeah dataclass -> dict conversion is tricky... I'll try to think if there's an opportunity to further optimize but I can't make any promises. Take a look at Dan's suggestions it may help you overcome this issue in a different way.

**Comment by danpovey** [comment]
There was a direction that I was hoping we could go in order to easily support very large datasets, which is to start using stateless samplers (that always select elements totally i.i.d. with no attempt to avoid duplicates), and the sampler would be initialized with a list of manifests; the sampling procedure would be to randomly select a manifest then randomly select an item.  I believe Piotr implemented this.  To make it efficient you need to create a file that tells you the start-char of each line in the json file; I believe this was done.  Then, there isn't a strong need to process the entire dataset into a single jsonl file; you can just have it be a list of jsonl files and they don't even have to be homogeneous in any sense.   I suspect all the tools to do this already exist although I forget the class names / PR numbers.  The only thing is, if you restart training you have to reset the seed to something that depends on the (restarted) batch number so that it doesn't use the exact same items as you used on batch 0.

The idea would be that generally speaking we wouldn't use the concept of epoch for training, but just the batch number; see the Eden2 learning rate schedule, which supports this, if you were previously using the (epoch-dependent) Eden.

**Reply by pzelasko** [reply]
The mechanism Dan mentioned is implemented here: https://github.com/lhotse-speech/lhotse/blob/d474be6b1d469863064142fc99e0861d164d8311/lhotse/dataset/sampling/stateless.py#L20
